### ISSUE TITLE
FR:  Additional configuration for k8s-operator spawned pods

### ISSUE THREAD

**Issue by ChandonPierre** [tailscale/tailscale#10880]
### What are you trying to do?

the k8s-operator creates a new `Statefulset` per exposed service - the operator allows configuration of some parameters via env var; e.g `PROXY_IMAGE` - however there are only a limited subset of configurable options. So far, I've had to handle these in a mutating webhook:

- tolerations
- node affinity
- requests/limits
- env vars; e.g `TS_EXTRA_ARGS` (need to add `--netfilter-mode=off` to avoid tailscale blackholing routes)
- container `postStart` lifecycle hook (for `iptables` related to the above netfilter argument)

This is not an exhaustive list of pod spec/statefulset spec that should be exposed and configurable

### How should we solve this?

Support for this can be added by exposing additional env vars (this can get out of hand pretty quickly), providing the operator a statefulset template in a configmap it then patches, or provide a json patch in a configmap the operator applies to the statefulset spec it generates

### What is the impact of not solving this?

Without a solve for this, end users will either have to deploy statefulsets manually (which largely defeats the utility of the operator)  or use mutating webhooks

**Comment by irbekrm** [comment]
Custom user-defined labels and other customization can now (v1.60) be applied via a ProxyClass custom resource https://tailscale.com/kb/1236/kubernetes-operator#cluster-resource-customization-using-proxyclass-custom-resource.

We are planning to eventually add more customization options to ProxyClass.
Feedback on whether it fits your workflow + what other fields you might need to configure (and why) is very welcome.

**Comment by irbekrm** [comment]
Hi @samcday , thank you for confirming that it solves your use case!

> would greatly benefit from a cluster-scoped (i.e non-namespaced) ClusterProxyClass

`ProxyClass` [is cluster scoped](https://github.com/tailscale/tailscale/blob/v1.63.0-pre/cmd/k8s-operator/deploy/crds/tailscale.com_proxyclasses.yaml#L14) . It should be possible to create one `ProxyClass` and refer to it from any proxy within any namespace - are you seeing issues with that?

We _did not_ name it `ClusterProxyClass` because we were not planning on making a namespace-scoped `ProxyClass`, so the `Cluster`- prefix seemed redundant.
Probably we should clarify that it's cluster-scoped in docs/CRD comments.

**Comment by samcday** [comment]
Haha! I'm not sure why I assumed it wasn't cluster scoped. My mistake :)

**Comment by irbekrm** [comment]
Hi @ChandonPierre , we did not expose the whole Pod spec via `ProxyClass` because if we did, we would inherit the whole of upstream validation rules and also because of the size of the CRD (4000 lines). Instead we opted for adding the fields folks need.

> We still mutate this lifecycle postStart on proxy pods in flight: while ! ip -f inet addr show tailscale0 &> /dev/null; do sleep 1; done; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE"

Why do you need this?

**Comment by irbekrm** [comment]
Makes sense, I'm generally happy for us to add a [`Lifecycle` field](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2776-L2779) to [the tailscale container](https://github.com/tailscale/tailscale/blob/main/k8s-operator/apis/v1alpha1/types_proxyclass.go#L113). If you would like to create a PR for this, please go ahead else I will put something in our backlog.

**Comment by irfn** [comment]
I need to have deployments/nameserver have some tolerations. This doesn't respect any default toleration config given.

I was able to get Operator, Proxy and egress pods to have the tolerations via mentioned setting like ProxyClass, Helm values at OperatorConfig etc.

Can this also be added to the list of configs. IMHO this is important as we often cant control what tolerations we need on a specific cluster.

**Comment by sinner-** [comment]
For anyone coming to this issue in 2026, the doc has moved here, it looks like: https://tailscale.com/docs/features/kubernetes-operator/how-to/customize#cluster-resource-customization-using-proxyclass
